File: androidHelpers/androidutils.cpp
#include "androidutils.h"
#include <QJniObject>
#include <QDebug>

QSize AndroidUtils::physicalScreenSize()
{
    QJniObject result = QJniObject::callStaticObjectMethod(
        "org/dsa/homepanel/AndroidUtils",
        "getRealScreenSize",
        "()Landroid/graphics/Point;"
        );

    if (!result.isValid()) {
        qWarning() << "AndroidUtils.getRealScreenSize() returned invalid";
        return {};
    }

    int width = result.getField<int>("x");
    int height = result.getField<int>("y");

    if (width <= 0 || height <= 0) {
        qWarning() << "Invalid screen size:" << width << "x" << height;
        return {};
    }

    qDebug() << "✅ Реальный размер экрана:" << width << "x" << height;
    return {width, height};
}

float AndroidUtils::physicalDensity()
{
    QJniObject metrics = QJniObject::callStaticObjectMethod(
        "org/dsa/homepanel/AndroidUtils",
        "getDisplayMetrics",
        "()Landroid/util/DisplayMetrics;"
        );

    if (!metrics.isValid()) {
        qWarning() << "AndroidUtils.getDisplayMetrics() returned invalid";
        return 1.0f;
    }

    float density = metrics.getField<jfloat>("density");
    qDebug() << "✅ Физическая плотность (density):" << density;
    return density;
}

File: android/src/org/dsa/homepanel/AndroidUtils.java
package org.dsa.homepanel;

import android.app.Activity;
import android.graphics.Point;
import android.util.DisplayMetrics;
import android.view.Display;

public class AndroidUtils {

    // Статическое хранение контекста активности
    private static Activity sActivity;

    public static void setActivity(Activity activity) {
        sActivity = activity;
    }

    // Получить реальные размеры экрана в пикселях
    public static Point getRealScreenSize() {
        if (sActivity == null) return new Point(-1, -1);

        Display display = sActivity.getWindowManager().getDefaultDisplay();
        Point size = new Point();
        display.getRealSize(size); // включает системные панели (если нет вырезов)
        return size;
    }

    // Альтернатива: через DisplayMetrics
    public static DisplayMetrics getDisplayMetrics() {
        if (sActivity == null) return null;

        DisplayMetrics metrics = new DisplayMetrics();
        sActivity.getWindowManager().getDefaultDisplay().getRealMetrics(metrics);
        return metrics;
    }
}
